File: server-rs/crates/cordy-middleware/src/plugin_auth.rs
//! Plugin auth routing — port of `server/internal/middleware/plugin_auth.go`.
//!
//! Lets the Action API be reached two ways without weakening either:
//!
//! The original way is the only one a SURFACE has: no credential at all. The
//! iframe posts a message to the host page, the host re-issues the call on the
//! signed-in user's session, and this middleware sends it through the ordinary
//! Auth chain like any other request.
//!
//! Hooks add a second way. A plugin's own server has no session and never
//! will, so when it presents a plugin bearer token this middleware steps
//! aside and lets the handler resolve the token itself — which it must,
//! because only the handler knows which installation and which scopes that
//! token stands for.
//!
//! Stepping aside is not the same as skipping authentication: every Action API
//! handler starts by resolving a caller, and a request that arrives with
//! neither a session nor a valid token fails there.

use axum::extract::{Request, State};
use axum::http::{header, HeaderMap, StatusCode};
use axum::middleware::Next;
use axum::response::Response;

use crate::auth::{auth_middleware, AuthState};

/// Pulls the raw credential out of an Authorization header.
/// Case-insensitive `Bearer ` prefix; empty when absent or malformed.
pub fn bearer_token(headers: &HeaderMap) -> String {
    let header_value = headers
        .get(header::AUTHORIZATION)
        .and_then(|v| v.to_str().ok())
        .unwrap_or("")
        .trim();
    if header_value.is_empty() {
        return String::new();
    }
    const PREFIX: &str = "Bearer ";
    if header_value.len() <= PREFIX.len()
        || !header_value[..PREFIX.len()].eq_ignore_ascii_case(PREFIX)
    {
        return String::new();
    }
    header_value[PREFIX.len()..].trim().to_string()
}

/// Reports whether a credential is one of ours to resolve.
///
/// Prefix-matched rather than validated: this only decides which code path
/// gets to look at the token, and an invalid token routed here is refused by
/// the handler a moment later. Deciding by prefix keeps a plugin token from
/// being tried against the PAT cache and a PAT from being tried against
/// installations.
pub fn is_plugin_bearer_token(token: &str) -> bool {
    token.starts_with("mpi_") || token.starts_with("mpc_")
}

/// PluginAuth wrapper — use via
/// `axum::middleware::from_fn_with_state(state, plugin_auth)`. Requests
/// bearing a plugin token go straight to the handler; everything else runs
/// through the ordinary Auth chain.
pub async fn plugin_auth(
    State(state): State<AuthState>,
    req: Request,
    next: Next,
) -> Result<Response, (StatusCode, &'static str)> {
    if is_plugin_bearer_token(&bearer_token(req.headers())) {
        return Ok(next.run(req).await);
    }
    auth_middleware(State(state), req, next).await
}

#[cfg(test)]
mod tests {
    use super::*;
    use axum::http::HeaderValue;

    fn hm(value: &str) -> HeaderMap {
        let mut m = HeaderMap::new();
        m.insert(header::AUTHORIZATION, HeaderValue::from_str(value).unwrap());
        m
    }

    #[test]
    fn bearer_token_extracts_case_insensitively() {
        assert_eq!(bearer_token(&hm("Bearer mpi_x")), "mpi_x");
        assert_eq!(bearer_token(&hm("bearer mpi_x")), "mpi_x");
        assert_eq!(bearer_token(&hm("BeArEr  mpi_x  ")), "mpi_x");
    }

    #[test]
    fn bearer_token_rejects_malformed() {
        assert_eq!(bearer_token(&HeaderMap::new()), "");
        assert_eq!(bearer_token(&hm("")), "");
        assert_eq!(bearer_token(&hm("Basic abc")), "");
        assert_eq!(bearer_token(&hm("Bearer")), "");
        assert_eq!(bearer_token(&hm("Bearer ")), "");
    }

    #[test]
    fn plugin_prefixes_recognized() {
        assert!(is_plugin_bearer_token("mpi_abc"));
        assert!(is_plugin_bearer_token("mpc_abc"));
        assert!(!is_plugin_bearer_token("mul_abc"));
        assert!(!is_plugin_bearer_token("mdt_abc"));
        assert!(!is_plugin_bearer_token(""));
    }
}

File: server/internal/middleware/plugin_auth.go
package middleware

import (
	"net/http"
	"strings"

	"github.com/patchbay-ai/patchbay/server/internal/auth"
	publicapiv1 "github.com/patchbay-ai/patchbay/server/pkg/publicapi/v1"
	db "github.com/patchbay-ai/patchbay/server/pkg/db/generated"
)

// PluginBearerOnly keeps a route on a machine-credential trust boundary.
// Prefix matching is only routing. The Action handler still resolves the token
// against its store and rejects an invalid or expired credential.
func PluginBearerOnly(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if !IsPluginBearerToken(BearerToken(r)) {
			publicapiv1.WriteProblem(w, r, http.StatusUnauthorized, "plugin_bearer_required", "plugin bearer token required")
			return
		}
		next.ServeHTTP(w, r)
	})
}

// PluginAuth lets the Action API be reached two ways without weakening either.
// Plugin bearer tokens skip the user session chain and are resolved by the
// handler. Every other request runs through ordinary Auth.
func PluginAuth(queries *db.Queries, patCache *auth.PATCache, cloudPAT *auth.CloudPATVerifier) func(http.Handler) http.Handler {
	userAuth := Auth(queries, patCache, cloudPAT)
	return func(next http.Handler) http.Handler {
		sessionHandler := userAuth(next)
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			if IsPluginBearerToken(BearerToken(r)) {
				next.ServeHTTP(w, r)
				return
			}
			sessionHandler.ServeHTTP(w, r)
		})
	}
}

// BearerToken pulls the raw credential out of an Authorization header.
func BearerToken(r *http.Request) string {
	header := strings.TrimSpace(r.Header.Get("Authorization"))
	if header == "" {
		return ""
	}
	const prefix = "Bearer "
	if len(header) <= len(prefix) || !strings.EqualFold(header[:len(prefix)], prefix) {
		return ""
	}
	return strings.TrimSpace(header[len(prefix):])
}

// IsPluginBearerToken reports whether a credential is one of ours to resolve.
//
// Prefix-matched rather than validated: this only decides which code path gets
// to look at the token, and an invalid token routed here is refused by the
// handler a moment later. Deciding by prefix keeps a plugin token from being
// tried against the PAT cache and a PAT from being tried against installations.
func IsPluginBearerToken(token string) bool {
	return strings.HasPrefix(token, "mpi_") || strings.HasPrefix(token, "mpc_")
}
